### PR TITLE
User defined additional kits

### PR DESCRIPTION
**Pull request by andreeis**
Inspired from the idea of https://github.com/microsoft/vscode-cmake-tools/pull/744.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1573,6 +1573,12 @@
             }
           }
         },
+        "cmake.additionalKitsFile": {
```
Comment (bobbrow) [inline package.json:1576]: >additionalKitsFile [](start = 15, length = 18)

Let's make it `additionalKitsFiles` and have it be array of string.

File: package.nls.json [hunk 2]
```diff
@@ -119,5 +119,6 @@
     "cmake-tools.configuration.cmake.statusbar.advanced.length.description": "Configures the maximum length of visible text in 'compact' mode.",
     "cmake-tools.configuration.views.cmake.folders.description": "Folders",
     "cmake-tools.configuration.views.cmake.outline.description": "Project Outline",
+    "cmake-tools.configuration.cmake.additionalKitsFile.description": "Path to a custom kit file",
```
Comment (bobbrow) [inline package.nls.json:122]: Update the description to match the array type

File: src/kitsController.ts [hunk 3]
```diff
@@ -46,17 +47,59 @@ export class KitsController {
...
     }
-    const kitsWatcher = chokidar.watch(KitsController._workspaceKitsPath(cmakeTools.folder), {ignoreInitial: true});
+
+    const folderKitsFiles: string[] = [KitsController._workspaceKitsPath(cmakeTools.folder)];
+    const additionalKitsFile: string | null = await KitsController.expandAdditionalKitsFile(cmakeTools);
+    if (additionalKitsFile) {
+      folderKitsFiles.push(additionalKitsFile);
+    }
+
+    const kitsWatcher = chokidar.watch(folderKitsFiles, {ignoreInitial: true});
```
Comment (bobbrow) [inline src/kitsController.ts:98]: Please double check this works with files outside the workspace folder

File: src/kitsController.ts [hunk 4]
```diff
@@ -119,17 +162,30 @@ export class KitsController {
...
-      // Read folder kits
+      // Read default folder kits
       this.folderKits = await readKitsFile(KitsController._workspaceKitsPath(this.folder));
-      const current = this.cmakeTools.activeKit;
-      if (current) {
-        const already_active_kit = this.availableKits.find(kit => kit.name === current.name);
-        // Set the current kit to the one we have named
-        await this.setFolderActiveKit(already_active_kit || null);
+
+      // Read additional folder kits
```
Comment (bobbrow) [inline src/kitsController.ts:170]: >// Read additional folder kits [](start = 6, length = 30)

Maybe the reading of kits can be factored out and reused by `init`

**Comment by andreeis** [comment]
The tests fail on clean develop as well, not sure what could have introduced this. The recent commits don't look like touching this area. Investigating...

File: package.json (hunk 1 above)
Comment (bobbrow) [reply package.json]: Please add `"Items"` property to enforce string for all array items.

---
In reply to: [587658754](https://github.com/Microsoft/vscode-cmake-tools/pull/1701#discussion_r587658754) [](ancestors = 587658754)

File: src/kitsController.ts [hunk 5]
```diff
@@ -46,17 +48,56 @@ export class KitsController {
...
+        userHome: paths.userHome,
+        workspaceFolder: cmakeTools.workspaceContext.folder.uri.fsPath,
+        workspaceFolderBasename: path.basename(cmakeTools.workspaceContext.folder.uri.fsPath),
+        workspaceHash: "",
+        workspaceRoot: cmakeTools.workspaceContext.folder.uri.fsPath,
+        workspaceRootFolderName: path.basename(cmakeTools.workspaceContext.folder.uri.fsPath),
+      }
+    };
+
+    await Promise.resolve(additionalKitFiles.forEach(async kitFile => {
```
Comment (bobbrow) [inline src/kitsController.ts:81]: We talked about using Promise.all here instead.  See `scanForVSKits` for an example.

File: package.json [hunk 6]
```diff
@@ -1631,6 +1631,12 @@
             }
           }
         },
+        "cmake.additionalKits": {
+          "type": "array",
+          "default": [],
```
Comment (bobbrow) [inline package.json:1636]: It looks like my other comment about adding the `"items"` property here was resolved by accident.  We still need to provide stronger validation here.

**Review by bobbrow — CHANGES_REQUESTED** [review]

File: package.json (hunk 6 above)
Comment (andreeis) [reply package.json:1636]: Oh right, I missed that.

**Review by bobbrow — APPROVED** [review]

**Pull request merged**